apps/src/gamelab/animationListModule.js

```diff
@@ -0,0 +1,605 @@
+        data: Object.assign({}, sourceAnimation, {
+          name: sourceAnimation.name + '_copy', // TODO: better generated names
+          version: versionId
+        })
+      });
+      dispatch(selectAnimation(newAnimationKey));
+      dashboard.project.projectChanged();
+    };
+
+    // If cloning a library animation, no need to perform a copy request
+    if (/^\/blockly\//.test(sourceAnimation.sourceUrl)) {
+      addClonedAnimation();
+    } else {
+      animationsApi.ajax(
+          'PUT',
+          newAnimationKey + '.png?src=' + key + '.png',
+          function success(xhr) {
+            try {
+              var response = JSON.parse(xhr.responseText);
+              addClonedAnimation(response.versionId);
+            } catch (e) {
+              onCloneError(e.message);
+            }
+          },
+          function error(xhr) {
+            onCloneError(xhr.status + ' ' + xhr.statusText);
+          });
+    }
+  };
+}
+
+/**
+ * Set the display name of the specified animation.
+ * @param {string} key
+ * @param {string} name
+ * @returns {{type: ActionType, key: string, name: string}}
+ */
+export function setAnimationName(key, name) {
+  return dispatch => {
+    dispatch({
+      type: SET_ANIMATION_NAME,
+      key,
+      name
+    });
+    dashboard.project.projectChanged();
+  };
+}
+
+/**
+ * Modifies the animation data, capturing changes to its spritesheet.
+ * @param {!AnimationKey} key
+ * @param {object} data - needs a more detailed shape
+ */
+export function editAnimation(key, data) {
+  return dispatch => {
+    dispatch({
+      type: EDIT_ANIMATION,
+      key,
+      data
+    });
+    dashboard.project.projectChanged();
+  };
+}
+
+/**
+ * Delete the specified animation from the project.
+ * @param {!AnimationKey} key
+ * @returns {function}
+ */
+export function deleteAnimation(key) {
+  return dispatch => {
+    animationsApi.ajax(
+        'DELETE',
+        key + '.png',
+        function success() {
+          dispatch(selectAnimation(null));
+          dispatch({
+            type: DELETE_ANIMATION,
+            key
+          });
+          dashboard.project.projectChanged();
+        },
+        function error(xhr) {
+          dispatch(reportError(`Error deleting object ${key}: ${xhr.status} ${xhr.statusText}`));
+        }
+    );
+  };
+}
+
+/**
+ * Load the indicated animation (which must already have an entry in the project
+ * animation list) from its source, whether that is S3 or the animation library.
+ * @param {!AnimationKey} key
+ * @param {string} [sourceUrl]
+ * @param {function} [callback]
+ */
+function loadAnimationFromSource(key, sourceUrl, callback) {
+  sourceUrl = sourceUrl || animationsApi.basePath(key) + '.png';
+  callback = callback || function () {};
+  return dispatch => {
+    dispatch({
+      type: START_LOADING_FROM_SOURCE,
+      key: key
+    });
+    // Figure out URL from animation key
+    // TODO: Take version ID into account here...
```

Comment:
TODO: This todo and comment belong at the top of the method, where we set the sourceUrl